syntax = "proto2";

package market;

// Dispersy messages
message UserMessage {
    required User user = 1;
}

message OfferMessage {
    optional LoanRequest loan_request = 1;
    optional Mortgage mortgage = 2;
    optional Investment investment = 3;
    optional Profile profile = 4;
    optional Transfer transfer = 5;
}

message AcceptMessage {
	required uint32 object_type = 1;
    required uint32 object_id = 2;
    required bytes object_user_id = 3;
}

message RejectMessage {
	required uint32 object_type = 1;
    required uint32 object_id = 2;
    required bytes object_user_id = 3;
}

message CampaignUpdateMessage {
    required Campaign campaign = 1;
    required Mortgage mortgage = 2;
    optional Investment investment = 3;
}


// Objects that are included in the Dispersy messages
message User {
    required bytes id = 1;
    required uint32 role = 2;
    optional Profile profile = 3;
}

message Profile {
    required string first_name = 1;
    required string last_name = 2;
    required string email = 3;
    required string iban = 4;
    required string phone_number = 5;
    optional string current_postal_code = 6;
    optional string current_house_number = 7;
    optional string current_address = 8;
    optional string document_list = 9;
}

message LoanRequest {
    required uint32 id = 1;
    required bytes user_id = 2;
    required House house = 3;
    required uint32 mortgage_type = 4;
    required bytes bank_id = 5;
    required string description = 6;
    required double amount_wanted = 7;
    required uint32 status = 8;
}

message House {
    required string postal_code = 1;
    required string house_number = 2;
    required string address = 3;
    required double price = 4;
    required string url = 5;
    required string seller_phone_number = 6;
    required string seller_email = 7;
}

message Mortgage {
    required uint32 id = 1;
    required bytes user_id = 2;
    required bytes bank_id = 3;
    required House house = 4;
    required double amount = 5;
    required double bank_amount = 6;
    required uint32 mortgage_type = 7;
    required double interest_rate = 8;
    required double max_invest_rate = 9;
    required double default_rate = 10;
    required uint32 duration = 11;
    required string risk = 12;
    required uint32 status = 13;
    optional Campaign campaign = 14;
    required uint32 loan_request_id = 15;
    required bytes loan_request_user_id = 16;
    optional bytes contract_id = 17;
}

message Investment {
    required uint32 id = 1;
    required bytes user_id = 2;
    required double amount = 3;
    required uint32 duration = 4;
    required double interest_rate = 5;
    required uint32 campaign_id = 6;
    required bytes campaign_user_id = 7;
    required uint32 status = 8;
    optional bytes contract_id = 9;
}

message Transfer {
    required uint32 id = 1;
    required bytes user_id = 2;
    required string iban = 3;
    required double amount = 4;
    required uint32 investment_id = 5;
    required bytes investment_user_id = 6;
    required uint32 status = 7;
    optional bytes contract_id = 8;
    optional bytes confirmation_contract_id = 9;
}

message Confirmation {
    required uint32 transfer_id = 1;
    required bytes transfer_user_id = 2;
    required string to_iban = 3;
    required string from_iban = 4;
    required double amount = 5;
}

message Campaign {
    required uint32 id = 1;
    required bytes user_id = 2;
    required uint32 mortgage_id = 3;
    required bytes mortgage_user_id = 4;
    required double amount = 5;
    required double amount_invested = 6;
    required uint32 end_time = 7;
}
